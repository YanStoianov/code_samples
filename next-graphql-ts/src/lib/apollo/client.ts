import { ApolloClient, InMemoryCache, HttpLink, ApolloLink, concat } from "@apollo/client";
import { CookieValueTypes, getCookie } from "cookies-next";
import { userConstants } from "@/store/user/constants";
import cacheOptions from "./cacheOptions";

interface ApolloClientProps {
  token?: CookieValueTypes;
}

const httpLink = new HttpLink({ uri: process.env.NEXT_PUBLIC_APOLLO_URI, includeExtensions: true });
const authToken = getCookie(userConstants.ACCESS_TOKEN);

const authMiddleware = ({ token }: ApolloClientProps) => {
  return new ApolloLink((operation, forward) => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    operation.setContext(({ headers }: Record<string, any>) => {
      return {
        headers: {
          authorization: token ? `Bearer ${token}` : "Bearer guest",
          ...headers,
        },
      };
    });
    return forward(operation).map((response) => {
      if (response.data) {
        response.data.extensions = response.extensions;
      }
      return response;
    });
  });
};

const client = new ApolloClient({
  ssrMode: true,
  link: concat(authMiddleware({ token: authToken }), httpLink),
  cache: new InMemoryCache(cacheOptions),
});

export const getApolloClient = ({ token }: ApolloClientProps) => {
  return new ApolloClient({
    ssrMode: true,
    link: concat(authMiddleware({ token }), httpLink),
    cache: new InMemoryCache(cacheOptions),
  });
};

export default client;
