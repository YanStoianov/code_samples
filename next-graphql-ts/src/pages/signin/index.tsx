import React from "react";
import { GetServerSideProps } from "next/types";
import useAuth, { AuthProps } from "@/hooks/useAuth";
import { getCookie } from "cookies-next";
import { useIntl } from "react-intl";
// COMPONENTS
import LogIn from "@/components/common/LogIn/LogIn";
import AuthLayout from "@/components/layouts/AuthLayout/AuthLayout";
import ApLoader from "@/components/loaders/ApLoader/ApLoader";
// CONSTANTS
import { userConstants } from "@/store/user/constants";

const SignIn = (props: AuthProps) => {
  const { accessToken } = props;
  const { showLoader, user } = useAuth({ accessToken });

  const intl = useIntl();

  const title = intl.formatMessage({ id: "auth.signIn" });

  if (showLoader || user) {
    return <ApLoader isVisible={true} />;
  }

  return (
    <AuthLayout title={title} description={"Apple | signin"}>
      <LogIn />
    </AuthLayout>
  );
};

export const getServerSideProps: GetServerSideProps = async (context) => {
  return {
    props: {
      accessToken: getCookie(userConstants.ACCESS_TOKEN, context) || null,
    },
  };
};

export default SignIn;
