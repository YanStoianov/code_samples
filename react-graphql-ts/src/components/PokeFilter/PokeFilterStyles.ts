import styled from "styled-components";

export default styled.div`
  .filters-block {
    display: flex;
    justify-content: center;
    align-items: center;
    padding: 20px 10px;

    .search {
      max-width: 300px;
      border-radius: 5px;
    }

    .types-select {
      min-width: 150px;
      max-width: 300px;
      margin: 0 10px;

      .ant-select-selector {
        border-radius: 5px;
      }
    }
  }

  .additional-cursor {
    display: flex;
    justify-content: center;
    align-items: center;
    padding: 10px 0;
    background: rgb(0 0 0 / 8%);

    .cursor-title {
      color: gray;
      padding: 0 10px;
    }

    span {
      font-size: 30px;
      color: #001528;
    }
  }

  .flex-row {
    display: flex;
    justify-content: center;
    flex-wrap: wrap;
  }

  .additional-filters-block {
    padding-bottom: 20px;
    background: rgb(0 0 0 / 8%);

    .newRow {
      width: 100%;
    }
  }
`;