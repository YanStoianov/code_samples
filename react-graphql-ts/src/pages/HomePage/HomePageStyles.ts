import styled from "styled-components";

export default styled.div`
  padding: 10px 30px;

  .pika-table {
    th {
      text-transform: capitalize;
    }

    tr td:first-child {
      text-transform: capitalize;
    }
  }

  .name-block {
    display: flex;
    justify-content: center;
    align-items: center;
    flex-direction: column;

    img {
      width: 100%;
    }
  }
`;