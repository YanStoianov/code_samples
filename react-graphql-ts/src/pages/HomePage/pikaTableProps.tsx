import {ColumnsType} from "antd/es/table";
import React from "react";
import {PokemonSpecy, PokemonStat, PokemonTypeGroup} from "../../interfaces/pokeInterfaces";
import {titleCase} from "../../utils/stringUtils";
import {CheckCircleOutlined, CloseCircleOutlined} from "@ant-design/icons";

export const generateStatColumns = (stats: string[]) => {
    return stats.map(stat => (
        {
            title: stat,
            dataIndex: 'pokemon_v2_pokemonstats',
            key: `pokemon_v2_pokemontypes${stat}`,
            width: '120px',
            sorter: (a: any, b: any) => {
                const statA = a.pokemon_v2_pokemonstats
                    .find((statItem: PokemonStat) => statItem.pokemon_v2_stat.name === stat)?.base_stat || 0
                const statB = b.pokemon_v2_pokemonstats
                    .find((statItem: PokemonStat) => statItem.pokemon_v2_stat.name === stat)?.base_stat || 0

                return statA - statB;
            },
            render: (statList: PokemonStat[]) => {
                const resultStat = statList.find((statItem) => statItem.pokemon_v2_stat.name === stat)
                return resultStat?.base_stat
            }
        }
    ))
}

interface DataType {
    key: React.Key;
    name: string;
    id: number;
    pokemon_v2_pokemontypes: [];
}


export const columns: ColumnsType<DataType> = [
    {
        title: 'Name',
        dataIndex: 'name',
        defaultSortOrder: 'ascend',
        width: '200px',
        sorter: (a, b) => a.name.localeCompare(b.name),
        render: (name, {id}) => {
            return (
                <div className={'name-block'}>
                    <img
                        src={`https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/${id}.png`}
                        alt={name}
                    />
                    <div>{name}</div>
                </div>
            );
        }
    },
    {
        title: 'Types',
        dataIndex: 'pokemon_v2_pokemontypes',
        width: '100px',
        render: (typeList) => {
            const types = typeList.map((item: PokemonTypeGroup) => {
                return item['pokemon_v2_type']['name'];
            }).join(',');
            return <div>{types}</div>;
        }
    },
    ...generateStatColumns(['hp', 'attack', 'defense', 'special-attack', 'special-defense', 'speed']),
    {
        title: 'Species',
        dataIndex: 'pokemon_v2_pokemonspecy',
        width: '450px',
        render: (species: PokemonSpecy) => {
            const speciesList: PokemonSpecy = {
                ...species,
                generation: species.pokemon_v2_generation?.name,
                growthrate: species.pokemon_v2_growthrate?.name
            };
            delete speciesList.pokemon_v2_growthrate;
            delete speciesList.pokemon_v2_generation;

            return <table className={'species-table'}>
                <tbody>
                {
                    Object.keys(speciesList).filter(key => key !== '__typename').map((key) => {
                        const value = (speciesList as any)[key];
                        let specyValue: JSX.Element = <>{JSON.stringify(value)}</>

                        if (value === true) {
                            specyValue = <CheckCircleOutlined className={'approved'}/>
                        }
                        if (value === false) {
                            specyValue = <CloseCircleOutlined className={'disapproved'}/>
                        }


                        return (
                            <tr key={key}>
                                <td className={'specy-name'}><b>{titleCase(key)}</b></td>
                                <td className={'specy-value'}>{specyValue}</td>
                            </tr>
                        )
                    })
                }
                </tbody>
            </table>
        }
    }
];