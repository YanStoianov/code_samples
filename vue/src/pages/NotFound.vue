<template>
  <div class="page-404">
    <h1>404</h1>
    <h2>Not Found</h2>
  </div>
</template>
<style scoped>
.page-404 {
  min-height      : 100%;
  min-width       : 100%;
  flex-grow       : 1;
  vertical-align  : middle;
  display         : flex;
  flex-direction  : column;
  align-items     : center;
  text-align      : center;
  justify-content : center;
}
</style>
