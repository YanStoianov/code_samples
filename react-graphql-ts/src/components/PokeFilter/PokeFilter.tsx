import {Input, Select} from "antd";
import React, {useMemo} from "react";
import PokeFilterStyles from "./PokeFilterStyles";
import {AdditionalPokeFilters, boolPokeFilters, minMaxPokeFilters, PokemonTypes} from "../../interfaces/pokeInterfaces";
import {DownSquareOutlined, UpSquareOutlined} from "@ant-design/icons";
import MinMaxInput from "../customInputs/MinMaxInput/MinMaxInput";
import {titleCase} from "../../utils/stringUtils";
import {initialAdditionalFilters} from "../../pages/HomePage";
import SwitchInput from "../customInputs/SwitchInput/SwitchInput";

const {Option} = Select;

interface PokeFilterProps {
    setNameSearch: (e: any) => void
    pokemonTypes: PokemonTypes | undefined
    setSelectedTypes: (value: number[]) => void
    setAdditionalFilters: (filters: AdditionalPokeFilters) => void
    additionalFilters: AdditionalPokeFilters
}

const PokeFilter = (props: PokeFilterProps) => {
    const {setNameSearch, pokemonTypes, setSelectedTypes, setAdditionalFilters, additionalFilters} = props;

    const typesSelect = useMemo(() => {
        return (
            <Select placeholder={'Types'} mode="multiple"
                    className={'types-select'} onChange={setSelectedTypes}>
                {
                    pokemonTypes?.pokemon_v2_type?.map(({id, name}) => <Option value={id} key={id}>{name}</Option>)
                }
            </Select>
        )
    }, [pokemonTypes, setSelectedTypes])

    const clearAdditionalFilters = () => {
        setAdditionalFilters({
            ...initialAdditionalFilters,
            useAdditionalFilters: false
        })
    }

    const setFilters = ({value, name}: any) => {
        setAdditionalFilters({
            ...additionalFilters,
            [name]: value
        })
    }

    const generateMinMaxFilters = (filterKeys: string[]) => {
        return filterKeys.map((key) => (
            <MinMaxInput
                key={key}
                title={titleCase(key)}
                setValue={(value) => setFilters({value, name: key})}
                // @ts-ignore
                value={additionalFilters[key]}
            />
        ))
    }

    const generateBoolFilters = (filterKeys: string[]) => {
        return filterKeys.map((key) => (
            <SwitchInput
                key={key}
                title={titleCase(key)}
                setValue={(value) => setFilters({value, name: key})}
                // @ts-ignore
                value={additionalFilters[key]}
            />
        ))
    }

    const {useAdditionalFilters} = additionalFilters;
    return (
        <PokeFilterStyles>
            <div className={'filters-block'}>
                <Input
                    placeholder="Search"
                    className={'search'}
                    onChange={e => setNameSearch(e.target.value)}
                />
                {typesSelect}
            </div>
            <div className={'additional-cursor'}>
                <div className={'cursor-title'}>Additional Filters</div>
                {
                    useAdditionalFilters ?
                        <UpSquareOutlined onClick={clearAdditionalFilters}/> :
                        <DownSquareOutlined onClick={() => setAdditionalFilters({
                            ...additionalFilters,
                            useAdditionalFilters: true
                        })}/>
                }
            </div>
            {useAdditionalFilters &&
                <div className={'additional-filters-block flex-row'}>
                    {generateMinMaxFilters(minMaxPokeFilters)}
                    <div className={'newRow flex-row'}>
                        {generateBoolFilters(boolPokeFilters)}
                    </div>
                </div>
            }
        </PokeFilterStyles>
    )
}

export default PokeFilter