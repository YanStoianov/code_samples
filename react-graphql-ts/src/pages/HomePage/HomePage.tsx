import React, {useMemo, useState} from 'react';
import MainLayout from '../../layouts/MainLayout';
import {useQuery} from "@apollo/client";
import {GET_POKEMON_TYPES, GET_POKEMONS} from "../../queries/pokQueries";
import {Table} from 'antd';
import LoaderWrapper from "../../components/LoaderWrapper/LoaderWrapper";
import HomePageStyles from "./HomePageStyles";
import {columns} from "./pikaTableProps";
import {
    AdditionalPokeFilters, boolPokeFilters,
    minMaxPokeFilters,
    PokemonInterface,
    PokemonTypes
} from "../../interfaces/pokeInterfaces";
import PokeStats from "../../components/PokeStats/PokeStats";
import PokeFilter from "../../components/PokeFilter/PokeFilter";
import inRange from "lodash/inRange";

const minMax = {
    min: 0,
    max: 0
}
export const initialAdditionalFilters: AdditionalPokeFilters = {
    base_happiness: {...minMax},
    capture_rate: {...minMax},
    gender_rate: {...minMax},
    hatch_counter: {...minMax},
    has_gender_differences: false,
    forms_switchable: false,
    is_legendary: false,
    is_mythical: false,
    generation: '',
    growthrate: '',
    is_baby: false,
    useAdditionalFilters: false
}

const HomePage: React.FC = () => {
    const [nameSearch, setNameSearch] = useState('');
    const [selectedTypes, setSelectedTypes] = useState<number[]>([]);
    const [additionalFilters, setAdditionalFilters] = useState<AdditionalPokeFilters>(initialAdditionalFilters);

    const {loading, data: pokemonData} = useQuery(GET_POKEMONS, {
        variables: {
            nameSearch: `%${nameSearch}%`
        }
    });
    const {data: pokemonTypes} = useQuery<PokemonTypes>(GET_POKEMON_TYPES);

    const tableData = useMemo(() => {
        const data = pokemonData?.pokemon_v2_pokemon || [];
        let result = data

        if (selectedTypes.length) {
            result = data.filter((pokemon: PokemonInterface) => {
                const pokemonTypes = pokemon.pokemon_v2_pokemontypes.map(({pokemon_v2_type}) => pokemon_v2_type.id);
                return pokemonTypes.some((typeId) => selectedTypes.includes(typeId))
            })
        }

        if (additionalFilters.useAdditionalFilters) {
            minMaxPokeFilters.forEach(filterName => {
                // @ts-ignore
                const filterValue = additionalFilters[filterName];

                if (filterValue.max <= 0) {
                    return null;
                }

                result = result.filter((pokemon: PokemonInterface) => {
                    // @ts-ignore
                    const pokemonValue = pokemon.pokemon_v2_pokemonspecy[filterName];
                    return inRange(pokemonValue, filterValue.min, filterValue.max + 1);
                })
            })
            boolPokeFilters.forEach(filterName => {
                // @ts-ignore
                const filterValue = additionalFilters[filterName];
                result = result.filter((pokemon: PokemonInterface) => {
                    // @ts-ignore
                    const pokemonValue = pokemon.pokemon_v2_pokemonspecy[filterName];
                    return pokemonValue === filterValue;
                })
            })
        }
        return result
    }, [pokemonData?.pokemon_v2_pokemon, selectedTypes, additionalFilters])

    return (
        <MainLayout>
            <HomePageStyles>
                <PokeStats pokemonTypes={pokemonTypes}/>
                <PokeFilter
                    setNameSearch={setNameSearch}
                    pokemonTypes={pokemonTypes}
                    setSelectedTypes={setSelectedTypes}
                    additionalFilters={additionalFilters}
                    setAdditionalFilters={setAdditionalFilters}
                />
                <LoaderWrapper loading={loading}>
                    <Table
                        className={'pika-table'}
                        columns={columns}
                        dataSource={tableData}
                        rowKey="name"
                    />
                </LoaderWrapper>
            </HomePageStyles>
        </MainLayout>
    )
};

export default HomePage;