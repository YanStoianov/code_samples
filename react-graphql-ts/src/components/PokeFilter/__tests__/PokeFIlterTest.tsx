import React from 'react'
import {render} from '@testing-library/react';
import 'jest-styled-components'
import PokeFilter from "../PokeFilter";
import {MockedProvider} from "@apollo/client/testing";
import {initialAdditionalFilters} from "../../../pages/HomePage";

describe("PokeFilterTest", () => {
    beforeAll(() => {
        Object.defineProperty(window, "matchMedia", {
            writable: true,
            value: jest.fn().mockImplementation(query => ({
                matches: false,
                media: query,
                onchange: null,
                addListener: jest.fn(),
                removeListener: jest.fn(),
                addEventListener: jest.fn(),
                removeEventListener: jest.fn(),
                dispatchEvent: jest.fn(),
            }))
        });
    });

    test('snapshot', () => {
        const {asFragment} = render(
            <MockedProvider addTypename={false}>
                <PokeFilter
                    setNameSearch={() => null}
                    pokemonTypes={undefined}
                    setSelectedTypes={() => null}
                    additionalFilters={initialAdditionalFilters}
                    setAdditionalFilters={() => null}
                />
            </MockedProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    })
});