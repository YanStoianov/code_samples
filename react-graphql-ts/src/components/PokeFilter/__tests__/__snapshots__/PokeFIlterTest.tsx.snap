// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokeFilterTest snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe piIVW"
  >
    <div
      class="filters-block"
    >
      <input
        class="ant-input search"
        placeholder="Search"
        type="text"
        value=""
      />
      <div
        class="ant-select types-select ant-select-multiple ant-select-show-search"
      >
        <div
          class="ant-select-selector"
        >
          <div
            class="ant-select-selection-overflow"
          >
            <div
              class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
              style="opacity: 1;"
            >
              <div
                class="ant-select-selection-search"
                style="width: 0px;"
              >
                <input
                  aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                  aria-autocomplete="list"
                  aria-controls="rc_select_TEST_OR_SSR_list"
                  aria-haspopup="listbox"
                  aria-owns="rc_select_TEST_OR_SSR_list"
                  autocomplete="off"
                  class="ant-select-selection-search-input"
                  id="rc_select_TEST_OR_SSR"
                  readonly=""
                  role="combobox"
                  style="opacity: 0;"
                  type="search"
                  unselectable="on"
                  value=""
                />
                <span
                  aria-hidden="true"
                  class="ant-select-selection-search-mirror"
                >
                  Â 
                </span>
              </div>
            </div>
          </div>
          <span
            class="ant-select-selection-placeholder"
          >
            Types
          </span>
        </div>
      </div>
    </div>
    <div
      class="additional-cursor"
    >
      <div
        class="cursor-title"
      >
        Additional Filters
      </div>
      <span
        aria-label="down-square"
        class="anticon anticon-down-square"
        role="img"
        tabindex="-1"
      >
        <svg
          aria-hidden="true"
          data-icon="down-square"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M505.5 658.7c3.2 4.4 9.7 4.4 12.9 0l178-246c3.8-5.3 0-12.7-6.5-12.7H643c-10.2 0-19.9 4.9-25.9 13.2L512 558.6 406.8 413.2c-6-8.3-15.6-13.2-25.9-13.2H334c-6.5 0-10.3 7.4-6.5 12.7l178 246z"
          />
          <path
            d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
          />
        </svg>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
